src/single_random_draw.rs

```diff
@@ -0,0 +1,450 @@
+//! This library provides efficient algorithms to compose a set of unspent transaction outputs
+//! (UTXOs).
+
+use crate::errors::Error;
+use crate::CHANGE_LOWER;
+use bitcoin::blockdata::transaction;
+use bitcoin::Amount;
+use bitcoin::FeeRate;
+use bitcoin::ScriptBuf;
+use bitcoin::Transaction;
+use bitcoin::TxIn;
+use bitcoin::TxOut;
+use bitcoin::Weight;
+use rand::seq::SliceRandom;
+
+/// Calculates the effective_value of an input.
+///
+/// Returns `Ok(None)` if the effective_value is negative.  If the effective_value is positive, return `Ok(Some(Amount))`.
+///
+/// ## Errors
+///
+/// Returns `Err(Error::Multiplication)` if `FeeRate` * `Weight` overflows.
+///
+fn get_effective_value(
+    utxo: &TxOut,
+    input: &TxIn,
+    fee_rate: FeeRate,
+) -> Result<Option<Amount>, Error> {
+    let weight_predict = transaction::InputWeightPrediction::new(
+        input.script_sig.len(),
+        input.witness.iter().map(|elem| elem.len()),
+    );
+
+    let input_weight = transaction::predict_weight(Some(vec![weight_predict]), None::<Vec<_>>);
+    let input_fee: Option<Amount> = fee_rate.checked_mul_by_weight(input_weight);
+
+    match input_fee {
+        Some(f) => Ok(utxo.value.checked_sub(f)),
+        None => Err(Error::Multiplication(input_weight, fee_rate)),
+    }
+}
+
+/// Calculate the fee for a transaction with two UTXOs and no inputs.
+///
+/// Returns `Ok(Amount)` where Amount is the fee.
+///
+/// ### Errors
+///
+/// Returns `Err(Error::Multiplication)` if `FeeRate` * `Weight` overflows.
+///
+fn calculate_fee(
+    script_pubkey: &ScriptBuf,
+    fee_rate: FeeRate,
+    transaction: &mut Transaction,
+) -> Result<Amount, Error> {
+    let output: TxOut = TxOut { value: Amount::ZERO, script_pubkey: script_pubkey.clone() };
+
+    let output_two = output.clone();
+
+    transaction.output = vec![output, output_two];
```

Comment:
1. It’s not clear to me why this assumes that both outputs have the same weight. What happens for example if your recipient output is an OP_RETURN output or a wrapped segwit output and your change is P2WPKH?
2. It’s not clear to me why this assumes always two outputs. What happens if you are trying to send all money in your wallet or to send to multiple recipients?

Reply:
this is outdated now.  Marking as resoled.